Asterix/optics/propagation_functions.py

```diff
@@ -416,3 +414,136 @@ def fft_choosecenter(image, inverse=False, center_pos='bb', norm='backward'):
+
+    Starting from the electric field in an input pupil plane, propagate through a phase-mask coronagraph with
+    different sampling values in different rings around the FPM center. Each propagation goes from the pupil to the FPM
+    and then to the Lyot plane (without Lyot stop applied), using MFTs, while a (combination of) Butterworth filters
+    marks annular regions around the FPM center that are considered in the propagation in each step. The electric field
+    contribution from each step is then summed up to yield the total electric field in the Lyot plane (before the LS).
+
+    AUTHOR: R. Galicher (in IDL)
+            ILa (to Python)
+
+    Parameters
+    ----------
+    pup : 2D array
+        Input mage array containing the wavefront at the entrance pupil of the optical system.
+    fpm : 2D array
+        Complex electric field in the focal plane of the focal-plane mask.
+    nbres : 1D array or list
+        List of the number of resolution elements in the total image plane for all propagation layers.
+    samp_outer : float
+        Sampling in the outermost layer of propagations.
+    filter_order : int
+        Order of the Butterworth filter.
+    alpha : float
+        Scale factor for the filter size. The larger this number, the smaller the filter size with respect to the
+        input array.
+
+    Returns
+    -------
+    array : E-field before the Lyot stop.
+    """
+    dim_pup = pup.shape[0]
+    dim_fpm = fpm.shape[0]
+
+    # Innermost part of the focal plane
+    but_inner = butterworth_circle(dim_fpm, dim_fpm / alpha, filter_order, -0.5, -0.5)
+    efield_before_fpm_inner = mft(pup, real_dim_input=dim_pup, dim_output=dim_fpm, nbres=nbres[0])
+    efield_before_ls = mft(efield_before_fpm_inner * fpm * but_inner, real_dim_input=dim_fpm, dim_output=dim_fpm,
+                           nbres=nbres[0], inverse=True)
+
+    # From inner to outer part of FPM
+    const_but = butterworth_circle(dim_fpm, dim_fpm / alpha, filter_order, xshift=-0.5, yshift=-0.5)
+    for k in range(nbres.shape[0] - 1):
+        # Butterworth filter in each layer
+        sizebut_here = dim_fpm / alpha * nbres[k] / nbres[k + 1]
+        but = (1 - butterworth_circle(dim_fpm, sizebut_here, filter_order, xshift=-0.5, yshift=-0.5)) * const_but
+
+        ef_pre_fpm = mft(pup, real_dim_input=dim_pup, dim_output=dim_fpm, nbres=nbres[k + 1])
+        ef_pre_ls = mft(ef_pre_fpm * fpm * but, real_dim_input=dim_fpm, dim_output=dim_fpm, nbres=nbres[k + 1],
+                        inverse=True)
+
+        # Sum up E-field contributions before the LS
+        efield_before_ls += ef_pre_ls
+
+    # Outer part of the FPM
+    nbres_outer = dim_fpm / samp_outer
+    sizebut_outer = dim_fpm / alpha * nbres[-1] / nbres_outer
+    but_outer = 1 - butterworth_circle(dim_fpm, sizebut_outer, filter_order, xshift=-0.5, yshift=-0.5)
+
+    ef_pre_fpm_outer = mft(pup, real_dim_input=dim_pup, dim_output=dim_fpm, nbres=nbres_outer, inverse=True)
```

Comment:
I can test this, but this would mean I could have to handle various cases of sampling, since the sampling of the outer plane is an input parameter for this function. Is this what we want?

Reply:
Hum, doing the last outer part at nyquist is always a good idea isnt ? This is waht Emiel does imo